vaadin-grid-flow-parent/vaadin-grid-flow/src/test/java/com/vaadin/flow/component/grid/GridTest.java

```diff
@@ -101,4 +109,66 @@ public void setAllRowsVisibleProperty_isHeightByRowsAndIsAllRowsVisibleWork() {
         Assert.assertTrue(grid.isHeightByRows());
         Assert.assertTrue(grid.isAllRowsVisible());
     }
+
+    @Test
+    public void dataProviderListeners_gridAttached_oldDataProviderListenerRemoved() {
+        AtomicInteger listenersCount = new AtomicInteger(0);
+        AtomicBoolean oldListenerRemoved = new AtomicBoolean();
+
+        // given
+        Grid<String> grid = new Grid<>();
+        ListDataProvider<String> dataProvider = new ListDataProvider<String>(
+                Collections.emptyList()) {
+            @Override
+            public Registration addDataProviderListener(
+                    DataProviderListener<String> listener) {
+                listenersCount.incrementAndGet();
+                Registration registration = super.addDataProviderListener(
+                        listener);
+                // the first listener is added by Grid in 'setDataProvider'
+                // and it is the one we want to be removed.
+                return listenersCount.get() == 1
```

Comment:
I'm trying to wrap my head around why is this needed for the test, but cannot, so please elaborate ? I understand collecting the number of listeners present, but don't understand why this changes the logic of the method directly. Seems odd.

Reply:
The goal of doing that is no only check that the given listener has been removed. Since there are several listeners added, including ones added by DataCommunicator, I use the known order of this listener. In case of Grid, the listener added in `setDataProvider` method by Grid is the first one in the list and it is supposed to be removed upon  `onAttach` method.